### PR TITLE
rcb for cartesian meshes

### PR DESCRIPTION
**Pull request by hhirtz**
Run the benchmark with

```
cargo bench --bench rcb_cartesian -- [--sample-size 10]
```

Run the test program with

```
cargo run -- WIDTH HEIGHT [RCB_ITERS]
```

### CODE REVIEW

File: src/cartesian.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+    fn position_of(&self, mut i: usize) -> [usize; D] {
+        let mut pos = [0; D];
+        for (s, p) in self.size.into_iter().zip(&mut pos) {
+            *p = i % s;
+            i /= s;
+        }
+        pos
+    }
+}
```
Comment (SebastienMorais) [inline src/cartesian.rs:29]: Grids are not necessarily squares (n x n) or cubes (n x n x n) but it seems to me that this hypothesis is leveraged in this function, no?

Comment (hhirtz) [reply src/cartesian.rs:29]: Is there something wrong I'm not seeing?

For a 2D grid of size `[W,H]`, the position of the `i`th element in a row-major array is `[i%W,I/W]`:

```
       <---- W ---->
^  0   1   2  ...  W-2  W-1
|  W  W+1 W+2 ... 2W-2 2W-1
H  *   *   *  ...   *   *
|  ........... i .........
v  *   *   *  ...   *  WH-1
```

For a 3D grid of size `[W,H,D]`, it's `[i%W,(i/W)%H,i/W/H]`.

The function computes the last component as `(i/W)%H` for 2D or `(i/W/H)%D` for 3D, but it doesn't change the result for `i`  smaller than `W*H` or `W*H*D`.

Comment (SebastienMorais) [reply src/cartesian.rs:29]: You are right i did misread the function :+1:

**Comment by hhirtz** [comment]
Grid size now use [`std::num::NonZeroUsize`][0] so that `Grid::position_of` cannot panic.

I also added 2D and 3D specializations for `Grid::position_of`:

image

becomes

image

also did it for `Grid::index_of`:

image

https://godbolt.org/z/PnnbP6Ef6

**Pull request merged**